src/Constraint/Constraint.php

```diff
@@ -255,22 +269,40 @@ public function matchSpecific(Constraint $provider, $compareBranches = false)
         // '!=' operator is match when other operator is not '==' operator or version is not match
         // these kinds of comparisons always have a solution
         if ($isNonEqualOp || $isProviderNonEqualOp) {
-            return (!$isEqualOp && !$isProviderEqualOp)
-                || $this->versionCompare($provider->version, $this->version, '!=', $compareBranches);
+            if ($isNonEqualOp && !$isProviderNonEqualOp && !$isProviderEqualOp && 'dev-' === substr($provider->version, 0, 4)) {
```

Comment:
we always call this `'dev-' === substr($provider->version, 0, 4)`  on ~all constraints.  We even performs it multiple time for the same constraint
What's about moving it in constructor?

Reply:
The bottleneck in Composer is often memory, so I prefer to opt for some extra computation over memory use a lot of the time. That said with it being the first thing versionCompare calculates, I guess it really does always get tested and branches already store a whole name as a string, so how much worse can adding a boolean flag really be. So guess it would make sense, yes.

Reply:
@jderusse Tried this out, I see no real performance gain and it costs about 1% increase in memory on a peak 100MB composer project. I've pushed the change here: https://github.com/naderman/semver/commit/499978306c90b38b820c7bd4413351115c6dcaf9 maybe you want to take a look at before/after on blackfire with the project(s) you checked before?